finality-aleph/src/sync/tasks.rs

```diff
@@ -0,0 +1,122 @@
+use std::{
+    collections::HashSet,
+    fmt::{Display, Error as FmtError, Formatter},
+    time::Duration,
+};
+
+use rand::{thread_rng, Rng};
+
+use crate::{
+    sync::{
+        data::BranchKnowledge,
+        forest::{Forest, Interest},
+        BlockIdFor, Header, Justification, PeerId,
+    },
+    BlockIdentifier,
+};
+
+const MIN_DELAY: Duration = Duration::from_millis(300);
+const ADDITIONAL_DELAY: Duration = Duration::from_millis(200);
+
+fn delay_for_attempt(attempt: u32) -> Duration {
+    MIN_DELAY
+        + ADDITIONAL_DELAY
+            .mul_f32(thread_rng().gen())
+            .saturating_mul(attempt)
+}
+
+/// A task for requesting blocks. Keeps track of how many times it was executed and what kind of
+/// request it is, highest justified or not.
+pub struct RequestTask<BI: BlockIdentifier> {
+    id: BI,
+    highest_justified: bool,
+    tries: u32,
+}
+
+impl<BI: BlockIdentifier> Display for RequestTask<BI> {
+    fn fmt(&self, f: &mut Formatter<'_>) -> Result<(), FmtError> {
+        match self.highest_justified {
+            true => write!(
+                f,
+                "highest justified request for {:?}, attempt {}",
+                self.id, self.tries
+            ),
+            false => write!(f, "block request for {:?}, attempt {}", self.id, self.tries),
+        }
+    }
+}
+
+type PreRequestFor<I, J> = (BlockIdFor<J>, BranchKnowledge<J>, HashSet<I>);
+type DelayedTask<BI> = (RequestTask<BI>, Duration);
```

Comment:
Hm, I think this requires a tad more work here and there to make sense...

Ok, so I worked on this a bit, and the first one is now a struct, because I found a decent excuse to make it into one. I couldn't really find such an excuse for the other type though – it's literally only there to be unpacked into two things after a call.